### ISSUE TITLE
False positive for no-const-assign?

### ISSUE THREAD

**Issue by damianobarbati** [eslint/eslint#10449]
Versions:
```
...
"eslintConfig": {
    "parser": "babel-eslint",
    "parserOptions": {
        "ecmaVersion": 2017,
        "sourceType": "module",
        "impliedStrict": true,
        "ecmaFeatures": {
            "jsx": true,
            "impliedStrict": true,
            "globalReturn": false,
            "experimentalObjectRestSpread": true
        }
    },
    "env": {
        "browser": true,
        "node": true,
        "jest": true,
        "es6": true
    },
    "plugins": [
        "react",
        "jest"
    ],
    "extends": [
        "eslint:recommended",
        "plugin:react/recommended",
        "plugin:jest/recommended"
    ],
    "rules": {
        "no-console": "off",
        "no-unreachable": "off",
        "no-unused-vars": "off",
        "no-extra-boolean-cast": "off"
    }
},
"devDependencies": {
    "babel-eslint": "^8.2.3",
    ...
    "eslint": "^4.19.1",
    "eslint-loader": "^2.0.0",
    "eslint-plugin-jest": "^21.15.0",
    "eslint-plugin-react": "^7.7.0",
    ...
}
```

Behaviour:
```
{
    async reauth ({ token, deviceId, version }) {
        const user = this.tokenToUser(token);
        const activityInterval = await this.getActivityInterval();
        Object.assign(ws, { id: user.id, password: user.password, token, deviceId, version, activityInterval } = {});
        return { ...user, token, activityInterval };
    },

    async getActivityInterval () {
        const activityInterval = await redis.get('activityInterval');
        return activityInterval || pkg.defaultActivityInterval;
    },
}
```
![image](https://user-images.githubusercontent.com/2225869/41084882-9fd418e4-6a35-11e8-86c2-0f3a8d736933.png)

**Comment by not-an-aardvark** [comment]
Hi,

I think this is working as intended because the expression `{ id: user.id, password: user.password, token, deviceId, version, activityInterval } = {}` is using a destructuring assignment to reassign the `token`, `deviceId`, `version`, and `activityInterval` variables, as well as the `user.id` and `user.password` properties.

What is the `= {}` on line 3 intended to do? It seems like this might be an error in the code.

**Comment by damianobarbati** [comment]
You are correct, that's to lend a default object for destructuring if none is provided.
My bad derived from a quick copy-paste from a function prototype!

Thanks and sorry for the noise :(
